File: Dome/dome/rules/xss.py
"""Cross-Site Scripting (XSS) detection rules."""
import re

PATTERNS = [
    # Basic script tag
    (r"(?i)<\s*script[^>]*>",                                        "XSS-001", "Script tag injection"),
    # Event handlers
    (r"(?i)\bon\w+\s*=\s*['\"]?.*?['\"]?",                           "XSS-002", "Inline event handler"),
    # javascript: URI
    (r"(?i)javascript\s*:",                                          "XSS-003", "javascript: URI scheme"),
    # vbscript: URI
    (r"(?i)vbscript\s*:",                                            "XSS-004", "vbscript: URI scheme"),
    # data: URI with script
    (r"(?i)data\s*:.*?(base64|text/html)",                           "XSS-005", "data: URI potential XSS"),
    # expression() (IE CSS)
    (r"(?i)expression\s*\(",                                         "XSS-006", "CSS expression()"),
    # document.cookie / document.write
    (r"(?i)document\.(cookie|write|location|domain)",                "XSS-007", "DOM manipulation"),
    # alert/confirm/prompt
    (r"(?i)\b(alert|confirm|prompt)\s*\(",                           "XSS-008", "JS dialog function"),
    # SVG with script/event
    (r"(?i)<\s*(svg|img|body|iframe|embed|object)[^>]*(on\w+|src\s*=\s*['\"]?javascript)", "XSS-009", "Tag+event/JS attribute"),
    # HTML entity obfuscation of <script
    (r"(?i)(&lt;|&#x3c;|&#60;)\s*script",                           "XSS-010", "Entity-encoded script tag"),
]

COMPILED = [(re.compile(p), rid, desc) for p, rid, desc in PATTERNS]


def check(value: str) -> list[dict]:
    hits = []
    for pattern, rule_id, description in COMPILED:
        if pattern.search(value):
            hits.append({"rule_id": rule_id, "description": description, "category": "xss"})
    return hits

File: Dome/dome-rs/src/engine/xss.rs
/// Cross-Site Scripting detection – operates on already-normalised input.
use once_cell::sync::Lazy;
use regex::Regex;
use super::Hit;

const SCORE: u32 = 20;

// (pattern, rule_id, description)
static RULES: Lazy<Vec<(Regex, &'static str, &'static str)>> = Lazy::new(|| {
    vec![
        (Regex::new(r"<\s*script[^>]*>").unwrap(),
         "XSS-001", "Script tag injection"),
        (Regex::new(r"\bon\w+\s*=\s*['\x22]?").unwrap(),
         "XSS-002", "Inline event handler (on*=)"),
        (Regex::new(r"javascript\s*:").unwrap(),
         "XSS-003", "javascript: URI"),
        (Regex::new(r"vbscript\s*:").unwrap(),
         "XSS-004", "vbscript: URI"),
        (Regex::new(r"data\s*:.*?(base64|text/html)").unwrap(),
         "XSS-005", "data: URI with HTML/script content"),
        (Regex::new(r"expression\s*\(").unwrap(),
         "XSS-006", "CSS expression()"),
        (Regex::new(r"document\.(cookie|write|location|domain)").unwrap(),
         "XSS-007", "DOM manipulation"),
        (Regex::new(r"\b(alert|confirm|prompt)\s*\(").unwrap(),
         "XSS-008", "JS dialog function"),
        (Regex::new(r"<\s*(svg|img|body|iframe|embed|object)[^>]*(on\w+|src\s*=\s*javascript)").unwrap(),
         "XSS-009", "Tag with event/JS attribute"),
        (Regex::new(r"</\s*script\s*>").unwrap(),
         "XSS-010", "Closing script tag"),
        (Regex::new(r"srcdoc\s*=").unwrap(),
         "XSS-011", "srcdoc iframe injection"),
        (Regex::new(r"\b(innerhtml|outerhtml)\s*=").unwrap(),
         "XSS-012", "innerHTML/outerHTML assignment"),
        (Regex::new(r"\beval\s*\(").unwrap(),
         "XSS-013", "eval() call"),
        (Regex::new(r"\$\{[^}]+\}").unwrap(),
         "XSS-014", "Template literal injection"),
    ]
});

pub fn check(normalised: &str) -> Vec<Hit> {
    RULES.iter().filter_map(|(re, id, desc)| {
        if re.is_match(normalised) {
            Some(Hit { rule_id: id, description: desc, category: "xss", score: SCORE })
        } else {
            None
        }
    }).collect()
}
